core/arch/x86/decode.c

```diff
@@ -708,6 +708,135 @@ read_vex(byte *pc, decode_info_t *di, byte instr_byte,
     return pc;
 }
 
+/* Given the potential first evex byte at pc, reads any subsequent evex
+* bytes (and any prefix bytes) and sets the appropriate prefix flags in di.
+* Sets info to the entry for the first opcode byte, and pc to point past
+* the first opcode byte.
+*/
+static byte *
+read_evex(byte *pc, decode_info_t *di, byte instr_byte,
+          const instr_info_t **ret_info INOUT)
+{
+
+    const instr_info_t *info;
+    byte prefix_byte = 0, evex_pp = 0;
+    ASSERT(ret_info != NULL && *ret_info != NULL);
+    info = *ret_info;
+
+    CLIENT_ASSERT(info->type == EVEX_PREFIX_EXT, "internal evex decoding error");
+
+    /* If 64-bit mode, P[3:2] are 0 and P[10] is 1 then this is evex */
+    if (X64_MODE(di) && !TEST(0xC, *pc) && TEST(0x04, *(pc + 1))) {
+        info = &evex_prefix_extensions[0][1];
+    } else {
+
+        CLIENT_ASSERT(!X64_MODE(di), "internal evex decoding error");
+
+        /* not evex */
+        *ret_info = &evex_prefix_extensions[0][0];
+        return pc;
+    }
+
+    CLIENT_ASSERT(info->code == PREFIX_EVEX, "internal evex decoding error");
+
+    /* read 2nd evex byte */
+    instr_byte = *pc;
+    prefix_byte = instr_byte;
+    pc++;
+
+    if (TESTANY(PREFIX_REX_ALL | PREFIX_LOCK, di->prefixes) || di->data_prefix ||
+        di->rep_prefix || di->repne_prefix) {
+        /* #UD if combined w/ EVEX prefix */
+        *ret_info = &invalid_instr;
+        return pc;
+    }
+
+    byte evex_mm;
+    CLIENT_ASSERT(info->type == PREFIX, "internal vex decoding error");
+    /* fields are: R, X, B, R', 00, mm.  R, X, B and R' are inverted. */
+    if (!TEST(0x80, prefix_byte)) {
+        di->prefixes |= PREFIX_REX_R;
+    }
+
+    if (!TEST(0x40, prefix_byte)) {
+        di->prefixes |= PREFIX_REX_X;
+    }
+
+    if (!TEST(0x20, prefix_byte)) {
+        di->prefixes |= PREFIX_REX_B;
+    }
+
+    if (!TEST(0x10, prefix_byte)) {
+        di->prefixes |= PREFIX_EVEX_RR;
+    }
+
+    evex_mm = instr_byte & 0x3;
+
+    if (evex_mm == 1) {
+        *ret_info = &escape_instr;
+    } else if (evex_mm == 2) {
+        *ret_info = &escape_38_instr;
+    } else if (evex_mm == 3) {
+        *ret_info = &escape_3a_instr;
+    } else {
+        /* #UD: reserved for future use */
+        *ret_info = &invalid_instr;
+        return pc;
+    }
+
+    /* read 3rd evex byte */
+    prefix_byte = *pc;
+    pc++;
+
+    /* fields are: W, vvvv, 1, PP */
+    if (TEST(0x80, prefix_byte)) {
+        di->prefixes |= PREFIX_REX_W;
+    }
+
+    evex_pp = prefix_byte & 0x03;
+    di->vex_vvvv = (prefix_byte & 0x78) >> 3;
```

Comment:
Rename to vvvv? Or better: make a union with evex_vvvv, so other code like operand decoder can later refer to by an unique name. This implies changes further below in this patch, I think the AVX-512 instructions need new table entries with new operand types (i.e. TYPE_B into TYPE_BB, unfortunately there are no predefined type name in the Intel manuals). I haven't thought this through, but can we use existing instruction tables with just wider operands? That can't be true, since there are now mask operands, i.e. different, more operands, etc.

Reply:
vex_vvvv and evex_vvvv have the same logic. I not understand why  rename vex_vvvv?
I agree with "AVX-512 instructions need new table entries with new operand types" point. I will add that table in my future pull requests, when I will support disassemble of AVX-512 instructions.